composer/2022_airflow_summit/data_analytics_process_expansion.py

```diff
@@ -0,0 +1,235 @@
+    # The tables will have the following format
+    # +------------------+-------------------+------------------+
+    # |PHX_PRCP/SNOW_1997|...                |PHX_PRCP/SNOW_2021|
+    # +------------------+-------------------+------------------+
+    # |DWA result (float)|...                |DWA result (float)|
+    # +------------------+-------------------+------------------+
+    phx_annual_prcp_df = spark.createDataFrame([[]], StructType([]))
+    phx_annual_snow_df = spark.createDataFrame([[]], StructType([]))
+
+    # Distance weighting algorithm (DWA)
+    def phx_dw_compute(input_list) -> float:
+        # Input_list is a list of Row object with format:
+        # [
+        #    ROW(ID='...', LATITUDE='...', LONGITUDE='...', YEAR='...', ANNUAL_PRCP/ANNUAL_SNOW='...'),
+        #    ROW(ID='...', LATITUDE='...', LONGITUDE='...', YEAR='...', ANNUAL_PRCP/ANNUAL_SNOW='...'),
+        #    ...
+        # ]
+
+        # Contains 1 / d^2 of each station
+        factor_list = []
+        # The total sum of 1 / d^2 of all the stations
+        factor_sum = 0.0
+        for row in input_list:
+            latitude = row[1]
+            longitude = row[2]
+            # Calculate the distance from each station to Phoenix
+            distance_to_phx = math.sqrt(
+                (phx_location[0] - latitude) ** 2 + (phx_location[1] - longitude) ** 2
+            )
+            # Calculate the distance factor of each station (1 / d^2)
+            distance_factor = 1.0 / (distance_to_phx**2)
+            factor_list.append(distance_factor)
+            factor_sum += distance_factor
+
+        # Contains the weights of each station
+        weights_list = []
+        for val in factor_list:
+            weights_list.append(val / factor_sum)
+
+        dwa_result = 0.0
+        for row in input_list:
+            # This is the annual prcipitation/snowfall of each station
+            annual_value = row[4]
+            # Weight of each station
+            weight = weights_list[input_list.index(row)]
+            dwa_result += weight * annual_value
+
+        return dwa_result
+
+    for year_val in range(1997, 2022):
+        # Collect() function returns a list of Row object.
+        # prcp_year and snow_year will be the input for the distance weighting algorithm
+        prcp_year = (
+            annual_df.where(
+                (annual_df.ELEMENT == "PRCP") & (annual_df.YEAR == year_val)
+            )
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_PRCP"))
+            .collect()
+        )
+
+        snow_year = (
+            annual_df.where(
+                (annual_df.ELEMENT == "SNOW") & (annual_df.YEAR == year_val)
+            )
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_SNOW"))
+            .collect()
+        )
+
+        phx_annual_prcp_df = phx_annual_prcp_df.withColumn(
```

Comment:
I had spoken to Kaiyang about this - it's not a huge deal, but it feels unintuitive to me to continuously append columns instead of adding new (YEAR, VAL) rows. What do you think?

Reply:
I will do whatever you think is best - whether that is leaving it as is or adding a new column - you tell me!

Reply:
Ok probably better then pivot it to be (YEAR, VAL).

```
YEAR | DISTANCE_WEIGHTING 
___________________________
1971 | <value>
```